File: cpp/Completer.cpp
// Copyright (C) 2011, 2012  Strahinja Val Markovic  <val@markovic.io>
//
// This file is part of YouCompleteMe.
//
// YouCompleteMe is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// YouCompleteMe is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with YouCompleteMe.  If not, see <http://www.gnu.org/licenses/>.

#include "standard.h"
#include "Completer.h"
#include "Utils.h"

#include <boost/bind.hpp>
#include <boost/make_shared.hpp>
#include <algorithm>

using boost::python::len;
using boost::python::extract;
using boost::packaged_task;
using boost::bind;
using boost::unique_future;
using boost::make_shared;
using boost::shared_ptr;
using boost::bind;
using boost::thread;

namespace YouCompleteMe
{

namespace
{

const unsigned int MAX_ASYNC_THREADS = 4;
const unsigned int MIN_ASYNC_THREADS = 2;

void ThreadMain( TaskStack &task_stack )
{
  while ( true )
  {
    ( *task_stack.Pop() )();
  }
}

} // unnamed namespace


Completer::Completer( const Pylist &candidates )
  : threading_enabled_( false )
{
  AddCandidatesToDatabase( candidates, "", "" );
}


Completer::Completer( const Pylist &candidates,
                      const std::string &filetype,
                      const std::string &filepath )
  : threading_enabled_( false )
{
  AddCandidatesToDatabase( candidates, filetype, filepath );
}


Completer::~Completer()
{
  foreach ( const CandidateRepository::value_type &pair,
            candidate_repository_ )
  {
    delete pair.second;
  }
}


// We need this mostly so that we can not use it in tests. Apparently the
// GoogleTest framework goes apeshit on us if we enable threads by default.
void Completer::EnableThreading()
{
  threading_enabled_ = true;
  InitThreads();
}


void Completer::AddCandidatesToDatabase( const Pylist &new_candidates,
                                         const std::string &filetype,
                                         const std::string &filepath )
{
  std::vector< Candidate *> &candidates =
    GetCandidateVector( filetype, filepath );

  int num_candidates = len( new_candidates );
  candidates.clear();
  candidates.reserve( num_candidates );

  std::string candidate_text;
  for (int i = 0; i < num_candidates; ++i)
  {
    candidate_text = extract< std::string >( new_candidates[ i ] );
    Candidate *&candidate = GetValueElseInsert( candidate_repository_,
                                                candidate_text, NULL );
    if ( !candidate )
      candidate = new Candidate( candidate_text );

    candidates.push_back( candidate );
  }
}


void Completer::CandidatesForQuery( const std::string &query,
                                    Pylist &candidates ) const
{
  CandidatesForQueryAndType( query, "", candidates );
}


void Completer::CandidatesForQueryAndType( const std::string &query,
                                           const std::string &filetype,
                                           Pylist &candidates ) const
{
  std::vector< Result > results;
  ResultsForQueryAndType( query, filetype, results );

  foreach ( const Result& result, results )
  {
    candidates.append( *result.Text() );
  }
}


Future Completer::CandidatesForQueryAndTypeAsync(
    const std::string &query,
    const std::string &filetype ) const
{
  // TODO: throw exception when threading is not enabled and this is called
  if (!threading_enabled_)
    return Future();

  // Try not to look at this too hard, it may burn your eyes.
  shared_ptr< packaged_task< AsyncResults > > task =
    make_shared< packaged_task< AsyncResults > >(
      bind( &Completer::ResultsForQueryAndType,
                   boost::cref( *this ),
                   query,
                   filetype ) );

  unique_future< AsyncResults > future = task->get_future();

  task_stack_.Push( task );
  return Future( move( future ) );
}

AsyncResults Completer::ResultsForQueryAndType(
    const std::string &query,
    const std::string &filetype ) const
{
  AsyncResults results = boost::make_shared< std::vector< Result > >();
  ResultsForQueryAndType( query, filetype, *results );
  return results;
}

void Completer::ResultsForQueryAndType( const std::string &query,
                                        const std::string &filetype,
                                        std::vector< Result > &results ) const
{
  FiletypeMap::const_iterator it = filetype_map_.find( filetype );
  if ( it == filetype_map_.end() )
    return;

  Bitset query_bitset = LetterBitsetFromString( query );

  foreach ( const FilepathToCandidates::value_type &path_and_candidates,
            *it->second )
  {
    foreach ( Candidate* candidate, *path_and_candidates.second )
    {
      if ( !candidate->MatchesQueryBitset( query_bitset ) )
        continue;

      Result result = candidate->QueryMatchResult( query );
      if ( result.IsSubsequence() )
        results.push_back( result );
    }
  }

  std::sort( results.begin(), results.end() );
}


std::vector< Candidate* >& Completer::GetCandidateVector(
    const std::string &filetype,
    const std::string &filepath )
{
  boost::shared_ptr< FilepathToCandidates > &path_to_candidates =
    filetype_map_[ filetype ];

  if ( !path_to_candidates )
    path_to_candidates.reset( new FilepathToCandidates() );

  boost::shared_ptr< std::vector< Candidate* > > &candidates =
    (*path_to_candidates)[ filepath ];

  if ( !candidates )
    candidates.reset( new std::vector< Candidate* >() );

  return *candidates;
}


void Completer::InitThreads()
{
  int threads_to_create =
    std::max( MIN_ASYNC_THREADS,
      std::min( MAX_ASYNC_THREADS, thread::hardware_concurrency() ) );

  for ( int i = 0; i < threads_to_create; ++i )
  {
    threads_.create_thread( bind( ThreadMain, boost::ref( task_stack_ ) ) );
  }
}


} // namespace YouCompleteMe

File: python/completers/completer.py
#!/usr/bin/env python
#
# Copyright (C) 2011, 2012  Strahinja Val Markovic  <val@markovic.io>
#
# This file is part of YouCompleteMe.
#
# YouCompleteMe is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# YouCompleteMe is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with YouCompleteMe.  If not, see <http://www.gnu.org/licenses/>.

import abc
import vim
import vimsupport
import ycm_core


class CompletionsCache( object ):
  def __init__( self ):
    self.line = -1
    self.column = -1
    self.raw_completions = []
    self.filtered_completions = []


  def CacheValid( self ):
    completion_line, _ = vimsupport.CurrentLineAndColumn()
    completion_column = int( vim.eval( "s:completion_start_column" ) )
    return completion_line == self.line and completion_column == self.column


class Completer( object ):
  __metaclass__ = abc.ABCMeta


  def __init__( self ):
    self.completions_future = None
    self.completions_cache = None


  def ShouldUseNow( self, start_column ):
    inner_says_yes = self.ShouldUseNowInner( start_column )
    previous_results_were_empty = ( self.completions_cache and
                                    not self.completions_cache.raw_completions )
    return inner_says_yes and not previous_results_were_empty


  def ShouldUseNowInner( self, start_column ):
    pass


  def CandidatesForQueryAsync( self, query ):
    if query and self.completions_cache and self.completions_cache.CacheValid():
      self.completions_cache.filtered_completions = (
        self.FilterAndSortCandidates(
          self.completions_cache.raw_completions,
          query ) )
    else:
      self.completions_cache = None
      self.CandidatesForQueryAsyncInner( query )


  def FilterAndSortCandidates( self, candidates, query ):
    if not candidates:
      return []

    if hasattr( candidates, 'words' ):
      candidates = candidates.words
    items_are_objects = 'word' in candidates[ 0 ]

    return ycm_core.FilterAndSortCandidates(
      candidates,
      'word' if items_are_objects else '',
      query )


  def CandidatesForQueryAsyncInner( self, query ):
    pass


  def AsyncCandidateRequestReady( self ):
    if self.completions_cache:
      return True
    else:
      return self.AsyncCandidateRequestReadyInner()


  def AsyncCandidateRequestReadyInner( self ):
    if not self.completions_future:
      # We return True so that the caller can extract the default value from the
      # future
      return True
    return self.completions_future.ResultsReady()


  def CandidatesFromStoredRequest( self ):
    if self.completions_cache:
      return self.completions_cache.filtered_completions
    else:
      self.completions_cache = CompletionsCache()
      self.completions_cache.raw_completions = self.CandidatesFromStoredRequestInner()
      self.completions_cache.line, _ = vimsupport.CurrentLineAndColumn()
      self.completions_cache.column = int(
        vim.eval( "s:completion_start_column" ) )
      return self.completions_cache.raw_completions


  def CandidatesFromStoredRequestInner( self ):
    if not self.completions_future:
      return []
    return self.completions_future.GetResults()


  def OnFileReadyToParse( self ):
    pass


  def OnCursorMovedInsertMode( self ):
    pass


  def OnCursorMovedNormalMode( self ):
    pass


  def OnBufferVisit( self ):
    pass


  def OnCursorHold( self ):
    pass


  def OnInsertLeave( self ):
    pass


  def OnCurrentIdentifierFinished( self ):
    pass


  def DiagnosticsForCurrentFileReady( self ):
    return False


  def GetDiagnosticsForCurrentFile( self ):
    return []


  def ShowDetailedDiagnostic( self ):
    pass

  def GettingCompletions( self ):
    return False


  @abc.abstractmethod
  def SupportedFiletypes( self ):
    pass


  def DebugInfo( self ):
    return ''
